languages/java/oso/src/test/java/com/osohq/oso/PolarTest.java

```diff
@@ -591,11 +601,102 @@ public void testIterators() throws Exception {
+    // GIVEN
     p.loadStr("f(x) if x > 2;");
+    // WHEN
+    List<HashMap<String, Object>> res = p.query("f(x)", true).results();
+    // THEN
+    assertEquals(1, res.size());
+    HashMap<String, Object> hm = res.get(0);
+    assertEquals(1, hm.size());
+    Object expr = hm.get("x");
+    assertEquals(
+        new Expression(
+            Operator.And, List.of(new Expression(Operator.Gt, List.of(new Variable("_this"), 2)))),
+        expr);
+  }
+
+  @Test
+  public void testPartialUnification() throws Exception {
+    // GIVEN
+    p.loadStr("f(x, y) if x = y;");
+    Variable x = new Variable("x");
+    Variable y = new Variable("y");
+
+    // WHEN
+    List<HashMap<String, Object>> results = p.queryRule("f", x, y).results();
+
+    // THEN
+    assertEquals(1, results.size());
+    HashMap<String, Object> bindings = results.get(0);
+    assertEquals(bindings.get("x"), y);
+    assertEquals(bindings.get("y"), x);
+  }
+
+  @Test
+  public void testPartial() {
+    // GIVEN
+    p.loadStr("f(1);");
+    p.loadStr("f(x) if x = 1 and x = 2;");
+
+    // WHEN
+    Predicate rule = new Predicate("f", List.of(new Variable("x")));
+    List<HashMap<String, Object>> results = p.query(rule, true).results();
+    assertEquals(1, results.size());
+    assertEquals(1, results.get(0).get("x"));
+
+    p.loadStr("g(x) if x.bar = 1 and x.baz = 2;");
+
+    Predicate gRule = new Predicate("g", List.of(new Variable("x")));
+    results = p.query(gRule, true).results();
+    assertEquals(1, results.size());
+    Expression expr = (Expression) results.get(0).get("x");
+    List<Object> args = (List<Object>) unwrapAnd(expr);
+    assertEquals(2, args.size());
+    assertEquals(
+        args.get(0),
+        new Expression(
+            Unify, List.of(1, new Expression(Dot, List.of(new Variable("_this"), "bar")))));
+    assertEquals(
+        args.get(1),
+        new Expression(
+            Unify, List.of(2, new Expression(Dot, List.of(new Variable("_this"), "baz")))));
+  }
+
+  private Object unwrapAnd(Expression expression) {
+    assertEquals(expression.getOperator(), Operator.And);
+    if (expression.getArgs().size() == 1) {
+      return expression.getArgs().get(0);
+    } else {
+      return expression.getArgs();
+    }
+  }
+
+  public static class User {}
 
-    Exception exception =
-        assertThrows(Exceptions.UnexpectedPolarTypeError.class, () -> p.query("f(x)").results());
-    assertTrue(exception.getMessage().contains("unbound"));
+  public static class Post {}
+
+  @Test
+  public void testPartialConstraint() {
+    p.registerClass(User.class, "User");
+    p.registerClass(Post.class, "Post");
+    p.loadStr("f(x: User) if x.user = 1;");
+    p.loadStr("f(x: Post) if x.post = 1;");
+    Variable x = new Variable("x");
+
+    Predicate rule = new Predicate("f", List.of(x));
+    List<HashMap<String, Object>> results =
+        p.query(rule, Map.of("x", new TypeConstraint(x, "User")), true).results();
```

Comment:
How is this working if ToPolar isn't implemented for it?

Reply:
It's possible that it's _not_ working, this might be the reason for 2 results instead of 1

Reply:
This is working now, I added the implementation to the PR. I noticed this test was different though as ported. The python test [expects 1 result](https://github.com/osohq/oso/blob/574c80b357ca2e2487c8d6442f86308ffd3e684f/languages/python/oso/tests/test_polar.py#L840).

Are there any other tests you changed while porting?

Reply:
Not to my knowledge. I erroneously thought this should return 2 results for some reason.

Reply:
Cool, I think it's just because the python test looks for two arguments in the And expression. I got a bit confused myself reading it.

I checked the other tests & they look good to me!